Contribute guides for TLS (#374)
kserve/website: docs/admin/serverless/servicemesh/README.md

}
```

## Turn on strict TLS on the entire service mesh {#mesh-wide-mtls}

> In the previous section, turning on strict mTLS on a namespace is discussed. Users requiring to lock down all workloads in the service mesh, Istio can be configured with [strict mTLS on the whole mesh](https://istio.io/latest/docs/tasks/security/authentication/mtls-migration/#lock-down-mutual-tls-for-the-entire-mesh).

Reviewer: ```suggestion
In the previous section, turning on strict mTLS on a namespace is discussed. Users requiring to lock down all workloads in the service mesh, Istio can be configured with [strict mTLS on the whole mesh](https://istio.io/latest/docs/tasks/security/authentication/mtls-migration/#lock-down-mutual-tls-for-the-entire-mesh).
```

Users requiring, or users are required?

Author: `For users requiring` .... missing `For` :-)

Author: ...either that, or change like:
Users requiring to lock down all workloads in the service mesh can configure Istio with strict mTLS on the whole mesh.
```

Not sure what's better...

Especially under a mesh-wide highly strict mTLS setup (involving both `PeerAuthentication` and `DestinationRule` resources as mentioned [above](#mesh-wide-mtls)), you will notice that workloads that are not part of the mesh will become blocked from calling InferenceServices. The easiest way to fix this is by [disabling the KServe top-level virtual service](#disable-top-level-vs). However, if you need the KServe top-level virtual service, you can still have a working hybrid environment by configuring KServe with its dedicated local Istio Gateway.

First, you will need to patch the Knative gateway, as shown [above for turning on mesh-wide strict mTLS](#mesh-wide-mtls). Then, you will need to create the following resources:

```yaml
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: kserve-local-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
    - hosts:
        - '*'
      port:
        name: http
        number: 8082
        protocol: HTTP
---
apiVersion: v1
kind: Service
metadata:
  name: kserve-local-gateway
  namespace: istio-system
  labels:

>     experimental.istio.io/disable-gateway-port-translation: "true"

Reviewer 2: What does this label do ? might be good to add an explanation

Author: I simply took it from Knative's gateway manifests, to have similar configurations here in KServe.
It is related to this issue: https://github.com/kubeflow/manifests/issues/2082. Istio added this label as a workaround.

AFAIK, Istio has already implemented the long-term solution, but Knative still hasn't moved to it. Istio project may simply drop this label, given it is experimental. I may propose the migration to Knative project, and then we could also move.

kind: Gateway
metadata:
  name: kserve-local-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
    - hosts:
        - '*'
      port:
        name: https
        number: 8445
        protocol: HTTPS
      tls:
        mode: SIMPLE
        credentialName: kserve-cluster-local-tls
---
apiVersion: v1
kind: Service
metadata:
  name: kserve-local-gateway
  namespace: istio-system
spec:
  ports:
    - name: https
      protocol: TCP
      port: 443
      targetPort: 8445
  selector:
    istio: ingressgateway
```

In the `istio-system` namespace, you will need to create a Secret named `kserve-cluster-local-tls` holding the TLS certificate. You may get this certificate from `cert-manager` operator.

Finally, edit the **ingress** config in the inferenceservice ConfigMap to have the following:

```bash
kubectl edit configmap/inferenceservice-config --namespace kserve

ingress : |- {
    "knativeLocalGatewayService": "knative-local-gateway.istio-system.svc.cluster.local",
    "localGateway":               "istio-system/kserve-local-gateway",
    "localGatewayService":        "kserve-local-gateway.istio-system.svc.cluster.local"

> }

Reviewer 2: Would be nice to include a diagram to explain the data plane path how the traffic is secured with these configured gateways

    <
    * Connection #0 to host sklearn-iris-burst-predictor-default.user1.svc.cluster.local left intact
    {"name":"sklearn-iris-burst","ready":true}
    ```

## Invoking InferenceServices from workloads that are not part of the mesh {#invoking-isvc-non-mesh}

> Ideally, when using service mesh, all involved workloads should belong to the service mesh. This allows enabling strict TLS in Istio and ensures policies are correctly applied. However, given the diverse requirements of applications, it is not always possible to migrate all workloads to the service mesh.

Reviewer: `all involved workloads should belong to the service mesh`

to the same service mesh?

Author: AFAIK, upstream Istio is a "singleton" on a cluster.

> Ideally, when using service mesh, all involved workloads should belong to the service mesh. This allows enabling strict TLS in Istio and ensures policies are correctly applied. However, given the diverse requirements of applications, it is not always possible to migrate all workloads to the service mesh.

Reviewer: `TLS` or `mTLS`?

Author: mTLS. I'll change.

> When using KServe, you may successfully migrate InferenceServices to the service mesh (i.e. by [injecting the Istio sidecar](#isvc-inject-sidecar)), while workloads invoking inferencing remain outside the mesh. In this hybrid environment, workloads that are not part of the service mesh need to use an Istio ingress gateway as port-of-entry to InferenceServices. In the default setup, KServe integrates with the gateway used by Knative. Both Knative and KServe apply the needed configurations to allow for these hybrid environments, however this only works transparently if you haven't enabled strict TLS on Istio.

Reviewer: ```suggestion
When using KServe, you may successfully migrate InferenceServices to the service mesh (i.e. by [injecting the Istio sidecar](#isvc-inject-sidecar)), while workloads invoking inferencing remain outside the mesh. In this hybrid environment, workloads that are not part of the service mesh need to use an Istio ingress gateway as a port-of-entry to InferenceServices. In the default setup, KServe integrates with the gateway used by Knative. Both Knative and KServe apply the needed configurations to allow for these hybrid environments, however, this only works transparently if you haven't enabled strict TLS on Istio.
```

> Specially under a mesh-wide strict mTLS setup, you will notice that workloads that are not part of the mesh will become blocked from calling InferenceServices. The easiest way to fix this is by [disabling the KServe top level virtual service](#disable-top-level-vs). However, if you need the KServe top level virtual service, you can still have a working hybrid environment by configuring KServe with its dedicated local Istio Gateway.

Reviewer: ```suggestion
Especially under a mesh-wide strict mTLS setup, you will notice that workloads that are not part of the mesh will become blocked from calling InferenceServices. The easiest way to fix this is by [disabling the KServe top-level virtual service](#disable-top-level-vs). However, if you need the KServe top-level virtual service, you can still have a working hybrid environment by configuring KServe with its dedicated local Istio Gateway.
```

> First, you will need to patch Knative gateway, as shown [above for turning on mesh-wide strict mTLS](#mesh-wide-mtls). Then, you will need to create the following resources:

Reviewer: ```suggestion
First, you will need to patch the Knative gateway, as shown [above for turning on mesh-wide strict mTLS](#mesh-wide-mtls). Then, you will need to create the following resources:
```

```yaml
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: kserve-local-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
    - hosts:
        - '*'
      port:
        name: http
        number: 8082
        protocol: HTTP
---
apiVersion: v1
kind: Service
metadata:
  name: kserve-local-gateway
  namespace: istio-system
spec:
  ports:
    - name: http
      protocol: TCP
      port: 80
      targetPort: 8081
  selector:
    istio: ingressgateway
```

> Finally, you need to edit **ingress** config in the inferenceservice ConfigMap to have the following:

Reviewer: ```suggestion
Finally, you need to edit the **ingress** config in the inference service ConfigMap to have the following:
```

```bash
kubectl edit configmap/inferenceservice-config --namespace kserve

ingress : |- {
    "knativeLocalGatewayService": "knative-local-gateway.istio-system.svc.cluster.local",
    "localGateway":               "istio-system/kserve-local-gateway",
    "localGatewayService":        "kserve-local-gateway.istio-system.svc.cluster.local"
}
```

After restarting KServe controller, the needed Services and VirtualServices will be reconfigured and workloads outside the mesh would be unblocked from calling InferenceServices.

## Securing the network border with TLS
If you require TLS over all the network, the easiest is to enable mesh-wide strict mTLS in Istio. Though, communications are secure only for workloads inside the service mesh and the mesh-border is the weak point: if the Istio ingress gateway is not setup with TLS, data arriving at or departing from the mesh would still be plain-text.

If you need the mesh-border to also be secured with TLS:

* For requests coming from outside the cluster (i.e. public endpoints), you can follow [Knative external domain encryption guide](https://knative.dev/docs/serving/encryption/external-domain-tls/)

> * For requests coming from workloads inside the cluster which are not part of the service mesh, you may use [Knative cluster local encryption](https://knative.dev/docs/serving/encryption/encryption-overview/#cluster-local-encryption), although it is still in experimental state. It would, also, require you to [disable the KServe top level virtual service](#disable-top-level-vs) to ensure there is no plain-text port-of-entry.

Reviewer: ```suggestion
* For requests coming from workloads inside the cluster that are not part of the service mesh, you may use [Knative cluster local encryption](https://knative.dev/docs/serving/encryption/encryption-overview/#cluster-local-encryption), although it is still in an experimental state. It would, also, require you to [disable the KServe top-level virtual service](#disable-top-level-vs) to ensure there is no plain-text port-of-entry.
```

If the experimental Knative cluster local encryption is not suitable for you, it is still possible to secure cluster-local traffic on the mesh border by configuring KServe with its own dedicated local gateway [similarly as shown previously](#invoking-isvc-non-mesh). The resources to create are as follows:

```yaml
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: kserve-local-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
    - hosts:

>         - '*'

Author: I saw in [Knative docs](https://knative.dev/docs/install/yaml-install/serving/install-serving-with-yaml/#__tabbed_1_2) that the provided manifests also use `*` for their local gateway. So, I simply used the same.

I can change if you think it is really better.

>         - '*'

Reviewer 2: This may need to match the cname in the tls credential

      port:
        name: https
        number: 8445
        protocol: HTTPS
      tls:
        mode: SIMPLE
        credentialName: kserve-cluster-local-tls
---
apiVersion: v1
kind: Service
metadata:
  name: kserve-local-gateway
  namespace: istio-system
spec:
  ports:
    - name: https
      protocol: TCP
      port: 443
      targetPort: 8445
  selector:
    istio: ingressgateway
```

> You will need to create a Secret named `kserve-cluster-local-tls` holding the TLS certificate. You may get this certificate from cert-manager.

Reviewer 3: It would be good to add here that the Secret should be created under the istio-system.

> You will need to create a Secret named `kserve-cluster-local-tls` holding the TLS certificate. You may get this certificate from cert-manager.

Reviewer: ```suggestion
You will need to create a Secret named `kserve-cluster-local-tls` holding the TLS certificate. You may get this certificate from `cert-manager` operator.
```

> Finally, edit **ingress** config in the inferenceservice ConfigMap to have the following:

Reviewer: ```suggestion
Finally, edit the **ingress** config in the inference service ConfigMap to have the following:
```

Author: I can add the suggested `the`, but I think `inferenceservice` is being used up in the other already existent sections. Would you like me to also change the others?

Reviewer: if there is a pattern already in the docs, we can follow it.

Author: OK, I added `the`, and left  `inferenceservice`.
